[AIRFLOW-910] Use parallel task execution in backfills

bolkedebruin opened the pull request:
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- AIRFLOW-910

The refactor to use dag runs in backfills caused a regression
in task execution performance as dag runs were executed
sequentially. Next to that, the backfills were non deterministic
due to the random execution of tasks, causing root tasks
being added to the non ready list too soon.

This updates the backfill logic as follows:
* Parallelize execution of tasks
* Use a leave first execution model; Breadth-first algorithm by Jerermiah
* Replace state updates from the executor by task based only updates

Will add tests after deemed acceptable.

@aoen @saguziel @mistercrunch @jlowin

ldct commented on airflow/models.py:
```diff
@@ -3032,6 +3032,25 @@ def get_task_instances(
     def roots(self):
         return [t for t in self.tasks if not t.downstream_list]
 
+    def topographical_sort(self):
```
Should be named topological sort (https://en.wikipedia.org/wiki/Topological_sorting) for consistency

jlowin replied:
@zodiac that's my fault :)

saguziel commented on airflow/models.py:
```diff
@@ -3032,6 +3032,25 @@ def get_task_instances(
     def roots(self):
         return [t for t in self.tasks if not t.downstream_list]
 
+    def topology_sort(self):
```
s/topology/topological

saguziel commented on airflow/models.py:
```diff
@@ -3032,6 +3032,25 @@ def get_task_instances(
 
+    def topology_sort(self):
+        """
+        Sorts tasks in topographical order, such that a task comes after any of its
```
topological

saguziel commented on airflow/models.py:
```diff
@@ -3032,6 +3032,25 @@ def get_task_instances(
+        Sorts tasks in topographical order, such that a task comes after any of its
+        upstream dependencies.
+        """
+        stack, visited, sort = [], set(), []
```
nit: s/sort/sorted_items or something similar

saguziel commented on airflow/models.py:
```diff
@@ -3032,6 +3032,25 @@ def get_task_instances(
+        stack, visited, sort = [], set(), []
+
+        # begin with any tasks that have
+        stack.extend(t for t in self.tasks if not t.upstream_list)
```
prefer explicitness with len(t.upstream_list) > 0

aoen commented on tests/models.py:
```diff
@@ -118,6 +118,33 @@ def test_dag_as_context_manager(self):
         self.assertEqual(dag.dag_id, 'creating_dag_in_cm')
         self.assertEqual(dag.tasks[0].task_id, 'op6')
 
+    def test_dag_topological_sort(self):
```
Seems like tests could be a bit more extensive here (e.g. test empty DAG).

bolkedebruin replied:
Will do

aoen commented on airflow/models.py:
```diff
@@ -3032,6 +3032,25 @@ def get_task_instances(
         return [t for t in self.tasks if not t.downstream_list]
 
+    def topological_sort(self):
+        """
```
Nit: Add return docstring to method signature

bolkedebruin replied:
Will do

aoen commented on airflow/models.py:
```diff
@@ -3032,6 +3032,25 @@ def get_task_instances(
+        Sorts tasks in topographical order, such that a task comes after any of its
+        upstream dependencies.
+        """
+        stack, visited, sorted_tasks = [], set(), []
```
Nit: don't think there is an advantage to combining these on the same line here

aoen commented on airflow/models.py:
```diff
@@ -3032,6 +3032,25 @@ def get_task_instances(
+        # begin with any tasks that have no upstream dependency
+        stack.extend(t for t in self.tasks if len(t.upstream_list) == 0)
+
+        while stack:
```
There are cases that this wouldn't work on (and we should add it to tests as well).

If a DAG looks like:
A->B
A->C->D
(where A -> B means task A depends on task B)
then your algorithm will have the stack looking like this over time:
B D
D A (INCORRECT! A should not be processed until C is processed).

Also visited is not being used here.

bolkedebruin replied:
I assume you say here A is dependent on both C and D? @jlowin as the "owner" of the algorithm would you mind taking a look?

aoen replied:
A depends on task on task C which depends on task D (A depends on D but only transitively)

bolkedebruin replied:
Gotcha. Good point.

aoen replied:
Here is an implementation I found online in case you want to use it:
http://blog.jupo.org/2012/04/06/topological-sorting-acyclic-directed-graphs/

bolkedebruin replied:
O nice! Will update tomorrow. Btw It seems we also have one in our JavaScript somewhere.

bolkedebruin committed:
[AIRFLOW-910] Use parallel task execution for backfills

    The refactor to use dag runs in backfills caused a regression
    in task execution performance as dag runs were executed
    sequentially. Next to that, the backfills were non deterministic
    due to the random execution of tasks, causing root tasks
    being added to the non ready list too soon.

    This updates the backfill logic as follows:
    * Parallelize execution of tasks
    * Use a leave first execution model; Breadth-first algorithm by Jerermiah
    * Replace state updates from the executor by task based only
